src/core/compact_object.cc

```diff
@@ -850,6 +863,20 @@ string_view CompactObj::GetSlice(string* scratch) const {
   return string_view{};
 }
 
+bool CompactObj::DefragIfNeeded(float ratio) {
+  // Supporting only for Strings - but not externals, or small string - they are using different
+  // allocations. otherwise we may crash See HasAllocated function for testing other conditions
+  const bool is_mimalloc_allocated =
+      (HasAllocated() && taglen_ != SMALL_TAG && ObjType() == OBJ_STRING);
```

Comment:
why logically it's correct, I suggest that you structure it differently so that when we add support for the SMALL_TAG it will be easily extended.

Reply:
I don't like it too, it was even more cumbersome. I think this is a classic case where you need a function.

Reply:
i meant separate tag_len_ test from others. and later write 
```
if (is_mimalloc_allocated) {
  if (taglen_ == ROBJ_TAG) {
    ...
  } 
  // else TODO
}
return false;
```

Reply:
yes - the next commit will have it, but since this can grow I think it would be more readable and less comments if this is in a function